pkg/rbac/common.go

```diff
@@ -0,0 +1,66 @@
+package rbac
+
+import (
+	"github.com/pkg/errors"
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	projectv3 "github.com/rancher/types/apis/project.cattle.io/v3"
+	rbacv1 "k8s.io/api/rbac/v1"
+)
+
+/*BuildSubjectFromObject Can accept PRTB, CRTB and APP as input.
+This function will generate PRTB and CRTB to the subject with user or group
+and generate APP to the subject with ServiceAccount
+*/
+func BuildSubjectFromObject(object interface{}) (rbacv1.Subject, error) {
+	var userName, groupPrincipalName, groupName, name, kind, namespace string
+	if rtb, ok := object.(*v3.ProjectRoleTemplateBinding); ok {
+		userName = rtb.UserName
+		groupPrincipalName = rtb.GroupPrincipalName
+		groupName = rtb.GroupName
+	} else if rtb, ok := object.(*v3.ClusterRoleTemplateBinding); ok {
+		userName = rtb.UserName
+		groupPrincipalName = rtb.GroupPrincipalName
+		groupName = rtb.GroupName
+	} else if app, ok := object.(*projectv3.App); ok {
+		userName = app.Name
+		namespace = app.Spec.TargetNamespace
+	} else {
+		return rbacv1.Subject{}, errors.Errorf("unrecognized roleTemplateBinding type: %v", object)
+	}
+
+	if userName != "" {
+		name = userName
+		kind = "User"
+	}
+
+	if groupPrincipalName != "" {
+		if name != "" {
+			return rbacv1.Subject{}, errors.Errorf("roletemplatebinding has more than one subject fields set: %v", object)
```

Comment:
the error message has a roletemplatebinding. And so do all the other messages.

Reply:
the comment is no longer relevant, as the method no longer handles an app object, only bindings.